File: cpp/2021/may/binary-tree-level-order-traversal.cc
#include <queue>
#include <vector>

#include "common.h"

using namespace std;

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;
        if (!root) {
            return result;
        }
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            vector<int> cur_level;
            auto        sz = q.size();
            for (decltype(sz) i = 0; i < sz; ++i) {
                TreeNode* node = q.front();
                q.pop();
                cur_level.push_back(node->val);
                if (node->left) {
                    q.push(node->left);
                }
                if (node->right) {
                    q.push(node->right);
                }
            }
            result.push_back(cur_level);
        }
        return result;
    }
};

File: src/tree/binary_tree_level_order_traversal.rs
use std::ops::Deref;

use super::OptNode;

use crate::Solution;

impl Solution {
    pub fn level_order(root: OptNode) -> Vec<Vec<i32>> {
        let mut queue = Vec::new();
        let mut ret = Vec::new();

        if let Some(r) = root {
            queue.push(r);
        }

        while !queue.is_empty() {
            let mut level = Vec::with_capacity(queue.len());
            let mut new_queue = Vec::with_capacity(queue.len() * 2);

            for node in queue.iter() {
                level.push(node.deref().borrow().val);
                if let Some(left) = node.borrow_mut().left.take() {
                    new_queue.push(left);
                }
                if let Some(right) = node.borrow_mut().right.take() {
                    new_queue.push(right);
                }
            }
            ret.push(level);
            queue = new_queue;
        }

        ret
    }
}
